g1ant-language/desktop-apps-automation2.md

```diff
@@ -0,0 +1,211 @@
+
+```G1
+if ⊂♥message⟦subject⟧.Contains("unsubscribe")⊃
+```
+
+When the condition is true (a matching message is found), the Robot will execute the next lines of code (in this exercise, call `deleteaddress` procedure) until `end if` is reached. Then the next message will be checked by the `foreach` command.
+
+In case the condition is false (the subject does not contain “unsubscribe”), the Robot should simply end the conditional loop with the `end if` command and move to the next message in a list.
+
+The full `checksubject` procedure code should read:
+
+```G1ANT
+procedure checksubject
+    foreach ♥message in ♥emails
+    	if ⊂♥message⟦subject⟧.Contains("unsubscribe")⊃
+            call deleteaddress
+        end if
+    end foreach
+end procedure
+```
+
+### Text Operations
+
+It’s time for the final part: the `deleteaddress` procedure, which will find the email address in the mailing list file and then delete it. The `text.` family of commands will help you do that.
+
+The information about the sender is stored in the `from` field of the mail structure. In other words, to retrieve it from the message, you use `⟦from⟧` index of the variable that stores the mail information — in this exercise it’s the `♥message` variable, so the correct expression is `♥message⟦from⟧`. The problem is that the `from` field includes not only email address (embraced in angle brackets), but also the sender’s name (if provided), for example: `John Doe <john.doe@email.org>` or `Company <office@company.com>`. Therefore, the first thing to do is to filter the email address out.
+
+You will use the `text.find` command for this purpose, feeding it with the text to be searched (sender information), the text to be searched for (everything within angle brackets) and the name of a variable that will store the result (the matching text — pure email address):
+
+```G1ANT
+procedure deleteaddress
+	text.find ♥message⟦from⟧ search ‴<✱>‴ result ♥address
+```
+
+Remember the asterisk called Search Place? You learned it in the [Basics](language-basics.md): it serves as a wildcard for all characters. You can enter this character from `Insert/Search Place` menu, with **Ctrl+8** keyboard shortcut or by clicking the `✱` icon on the toolbar.
+
+Now use the new `♥address` variable containing email address to search your mailing list file and delete this address. There is the `text.replace` command, which will do the job perfectly, but needs some additional action. Why? Because all `text.` commands operate on text — either provided directly to a command (e.g. `text.find ‴source text to be searched‴`) or indirectly with a variable (`text.find ♥searchedtext`) — not on a file containing text. If you want to search your mailing list file, you have to read its content first and store it in a variable. And here comes another `text.` command, which does exactly that: `text.read`.
```

Comment:
totally don't understand what the issue is :D

Reply:
"Because all `text.` commands operate on text — either provided directly to a command (e.g. `text.find ‴source text to be searched‴`) or indirectly with a variable (`text.find ♥searchedtext`)"
not only text. commands can take variables as argument values and I didn't see in your previous topics that you explained it for every command that you describe in the manual so I don't know why you do it here (I think it's quite obvious)

Reply:
It's not obvious that a specified command CAN'T use a FILE as an argument. I had to explain that text. commands don't operate on files, but one of them does and provides input for others. Got it now?

Reply:
No, I still don't get it. Now I don't even know what the whole sentence means. What do you mean by "command can't use a file as an argument" what is a file? a path to a file?

And why are you telling users that this particular set of commands can take variables as argument values? You have never explained it in other commands' descriptions so why here? This is what I found obvious. (Arguments can take variables storing the values to be passed to those arguments or just direct values.)

Reply:
Read this sentence again:
"Because all `text.` commands operate on text — either provided directly to a command [---] or indirectly with a variable [---] — not on a file containing text."
We assume that a reader doesn't know the command nor its arguments. So when you explain how the command works, you give as much information as you can. A user doesn't know what it means "to operate on text": what is text? Where does it come from? If you write it's given directly, you can't expect that a reader will figure out that you can also use a variable containing text. And vice Versace. ;)
Once again: what seems obvious to you, can be a total mystery for a rookie.

Reply:
"And why are you telling users that this particular set of commands can take variables as argument values? You have never explained it in other commands' descriptions so why here?"
You can leave it here but then it should be in every command' description right? So that the manual would be more consistent.

What I mean is: `Every command argument can take variables. Not only this one.`